src/extension/token/voting/VotingERC721.sol

```diff
@@ -0,0 +1,297 @@
+        _moveDelegateVotes(oldDelegate, delegatee, _getVotingUnits(account));
+    }
+
+    /**
+     * @notice Transfers, mints, or burns voting units. To register a mint, `from` should be zero. To register a burn, `to`
+     * should be zero. Total supply of voting units will be adjusted with mints and burns.
+     */
+    function _transferVotingUnits(address from, address to, uint256 amount) internal virtual {
+        Historical.Timeline storage totalVotes = _votingStorage().totalVotes;
+        if (from == address(0)) {
+            _push(totalVotes, _add, SafeCastLib.toUint208(amount));
+        }
+        if (to == address(0)) {
+            _push(totalVotes, _subtract, SafeCastLib.toUint208(amount));
+        }
+        _moveDelegateVotes(delegates(from), delegates(to), amount);
+    }
+
+    /// @notice Moves delegated votes from one delegate to another.
+    function _moveDelegateVotes(address from, address to, uint256 amount) private {
+        if (from != to && amount > 0) {
+            mapping(address => Historical.Timeline) storage delegatedVotes = _votingStorage().delegatedVotes;
+            if (from != address(0)) {
+                (uint256 oldValue, uint256 newValue) =
+                    _push(delegatedVotes[from], _subtract, SafeCastLib.toUint208(amount));
+                emit DelegateVotesChanged(from, oldValue, newValue);
+            }
+            if (to != address(0)) {
+                (uint256 oldValue, uint256 newValue) = _push(delegatedVotes[to], _add, SafeCastLib.toUint208(amount));
+                emit DelegateVotesChanged(to, oldValue, newValue);
+            }
+        }
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            UTIL FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    function _push(
+        Historical.Timeline storage timeline,
+        function(uint208, uint208) view returns (uint208) op,
+        uint208 delta
+    ) private returns (uint208, uint208) {
+        return timeline.push(clock(), op(timeline.latest(), delta));
+    }
+
+    function _add(uint208 a, uint208 b) internal pure returns (uint208) {
+        return a + b;
+    }
+
+    function _subtract(uint208 a, uint208 b) internal pure returns (uint208) {
+        return a - b;
+    }
+
+    /// @notice Returns the balance of `account`.
+    function _getVotingUnits(address account) internal view returns (uint256) {
+        return ERC721A(address(this)).balanceOf(account);
+    }
+
+    function _useNonce(address signer, uint256 nonce) internal {
+        if (nonce != _votingStorage().nonces[msg.sender]) revert InvalidNonce(nonce);
+        unchecked {
+            _votingStorage().nonces[signer] += 1;
+        }
+    }
+
+    /// @notice Returns the domain name and version for EIP712.
+    function _domainNameAndVersion() internal pure override returns (string memory name, string memory version) {
```

Comment:
how does EIP712 works in the context of an extension?

Core: EIP712 with the name of `ERC721Core`
Extension A installed on the Core: EIP712 with the name of `VotingERC721`
Extension B installed on the Core: EIP712 with the name of `RoyaltyERC721`

What is the signature the user is signing?
EIP712 by default includes contractAddress in the signature. In this specific case, it looks like user will need to sign signature for the extension implementation contract address instead of the core contract address. Which, could lead to signature-replay issue due to same extension contract signature can be used across different core contract because the execution happens through delegatecall

Reply:
Looked more into this and the `_hashTypedData` function in solady's EIP-712 catches for the signature-replay issue for when `delegateCall` is used from a different contract.

```
    function _hashTypedData(
        bytes32 structHash
    ) internal view virtual returns (bytes32 digest) {
        // We will use `digest` to store the domain separator to save a bit of gas.
        if (_domainNameAndVersionMayChange()) {
            digest = _buildDomainSeparator();
        } else {
            digest = _cachedDomainSeparator;
            if (_cachedDomainSeparatorInvalidated())
                digest = _buildDomainSeparator();
        }
        /// @solidity memory-safe-assembly
        assembly {
            // Compute the digest.
            mstore(0x00, 0x1901000000000000) // Store "\x19\x01".
            mstore(0x1a, digest) // Store the domain separator.
            mstore(0x3a, structHash) // Store the struct hash.
            digest := keccak256(0x18, 0x42)
            // Restore the part of the free memory slot that was overwritten.
            mstore(0x3a, 0)
        }
    }
```

Inside of the `else` statement block, the function `_cachedDomainSeparatorInvalidated` checks if `address(this)` is the same as the address used in the cached domain separator. In this case, it will fail since the cached address is the extension contract's address and `address(this)` will refer to the core contact's address

```
    function _cachedDomainSeparatorInvalidated()
        private
        view
        returns (bool result)
    {
        uint256 cachedChainId = _cachedChainId;
        uint256 cachedThis = _cachedThis;
        /// @solidity memory-safe-assembly
        assembly {
            result := iszero(
                and(eq(chainid(), cachedChainId), eq(address(), cachedThis))
            )
        }
    }
```

due to this, it will rebuild the domain separator from scratch, using the core contract's address instead, which should resolve this issue.

To save on gas, we can override the `_domainNameAndVersionMayChange` in the inheriting extension contracts so that it always returns true